src/fu-engine.c

```diff
@@ -6194,19 +6196,146 @@ fu_engine_record_security_attrs(FuEngine *self, GError **error)
+{
+	JsonArray *array;
+	JsonObject *obj;
+
+	/* sanity check */
+	if (!JSON_NODE_HOLDS_OBJECT(json_node)) {
+		g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "not JSON object");
+		return FALSE;
+	}
+
+	/* not supplied */
+	obj = json_node_get_object(json_node);
+	if (!json_object_has_member(obj, "Devices"))
+		return TRUE;
+
+	/* this has to exist */
+	array = json_object_get_array_member(obj, "Devices");
+	for (guint i = 0; i < json_array_get_length(array); i++) {
+		JsonNode *node_tmp = json_array_get_element(array, i);
+		g_autoptr(FuDevice) device = fu_device_new(self->ctx);
+		if (!fwupd_device_from_json(FWUPD_DEVICE(device), node_tmp, error))
+			return FALSE;
+		fu_device_set_plugin(device, "dummy");
+		fu_device_add_problem(device, FWUPD_DEVICE_PROBLEM_IS_EMULATED);
+		if (!fu_device_setup(device, error))
+			return FALSE;
+		fu_engine_add_device(self, device);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_engine_load_host_emulation(FuEngine *self, const gchar *fn, GError **error)
+{
+	g_autoptr(JsonParser) parser = json_parser_new();
+	g_autoptr(GFile) file = g_file_new_for_path(fn);
+	g_autoptr(GInputStream) istream_json = NULL;
+	g_autoptr(GInputStream) istream_raw = NULL;
+	g_autoptr(FwupdSecurityAttr) attr = NULL;
+
+	/* add an attr so we know this is emulated and do not offer to upload results */
+	attr = fwupd_security_attr_new(FWUPD_SECURITY_ATTR_ID_HOST_EMULATION);
+	fwupd_security_attr_set_plugin(attr, "core");
+	fwupd_security_attr_add_flag(attr, FWUPD_SECURITY_ATTR_FLAG_RUNTIME_ISSUE);
+	fwupd_security_attr_set_result(attr, FWUPD_SECURITY_ATTR_RESULT_ENABLED);
+	fu_security_attrs_append(self->host_security_attrs, attr);
+
+	/* add from file */
+	istream_raw = G_INPUT_STREAM(g_file_read(file, NULL, error));
+	if (istream_raw == NULL)
+		return FALSE;
+	if (g_str_has_suffix(fn, ".gz")) {
+		g_autoptr(GConverter) conv =
+		    G_CONVERTER(g_zlib_decompressor_new(G_ZLIB_COMPRESSOR_FORMAT_GZIP));
+		istream_json = g_converter_input_stream_new(istream_raw, conv);
+	} else {
+		istream_json = g_object_ref(istream_raw);
+	}
+	if (!json_parser_load_from_stream(parser, istream_json, NULL, error))
+		return FALSE;
+	if (!fu_engine_devices_from_json(self, json_parser_get_root(parser), error))
+		return FALSE;
+	if (!fu_engine_security_attrs_from_json(self, json_parser_get_root(parser), error))
+		return FALSE;
+
+#ifdef HAVE_HSI
+	/* depsolve */
+	fu_engine_security_attrs_depsolve(self);
+#endif
+
+	/* success */
+	return TRUE;
+}
+
 static void
 fu_engine_ensure_security_attrs(FuEngine *self)
 {
 #ifdef HAVE_HSI
 	GPtrArray *plugins = fu_plugin_list_get_all(self->plugin_list);
 	g_autoptr(GPtrArray) devices = fu_device_list_get_all(self->device_list);
-	g_autoptr(GPtrArray) items = NULL;
 	g_autoptr(GError) error = NULL;
 
 	/* already valid */
-	if (self->host_security_id != NULL)
+	if (self->host_security_id != NULL || self->host_emulation)
```

Comment:
I'm not convinced you want to avoid all this.  Don't you want the engine to do depsolve on all the attributes?  The way that i'm thinking about emulation is that you skip certain vfuncs in plugins that probe the device or probe the attributes and that you avoid letting you call write vfuncs.  So you don't avoid  as much engine functionality tied to all of the corner case behavior.

Reply:
Depsolve should be being done already. I'd like to keep it simple for now, and either have a 100% emulated host or a 100% real host -- otherwise it gets really complicated fast with things like custom FuDevice GTypes.